Add version field in PrivateStateToken  dictionary

aykutbulut opened the pull request:
Add version field in PrivateStateToken  dictionary. Fixes #136.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aykutbulut/trust-token-api/pull/209.html" title="Last updated on Mar 9, 2023, 4:33 AM UTC (ba74165)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/trust-token-api/209/4359cfb...aykutbulut:ba74165.html" title="Last updated on Mar 9, 2023, 4:33 AM UTC (ba74165)">Diff</a>

miketaylr commented:
@aykutbulut looks like you have an error in your IDL syntax, see https://github.com/WICG/trust-token-api/actions/runs/4285117949/jobs/7463967346#step:3:792

aykutbulut commented:
Pushed the fix.

johannhof reviewed (commented):
This seems fine in itself, but the spec algorithms also need to use this value, i.e. in the fetch steps we should return if the token version isn't `1`.

aykutbulut commented:
Updated issue and redemption request header algorithms. Let me know if some of the steps I added are redundant. 

For example, version field is required in PrivateStateToken dictionary. Does this mean a "TypeError" DOMException is implied if it is missing when constructing the Request?

johannhof commented on spec.bs:
```diff
@@ -405,6 +416,12 @@ fetch(issueRequest);
 
  1. If |request|'s [=request/client=] is not a [=secure context=], return.
+ 1. If |type| is not specified, throw a "{{TypeError}}" {{DOMException}}.
+ 1. If |type| is not `"private-state-token"`, throw a "{{TypeError}}" {{DOMException}}.
```
```suggestion
 1. If |type| is not `"private-state-token"`, throw a {{TypeError}}.
```

TypeError is a simple exception, see https://webidl.spec.whatwg.org/#idl-exceptions

johannhof replied:
(same below)

johannhof commented on spec.bs:
```diff
@@ -405,6 +416,12 @@ fetch(issueRequest);
+ 1. If |version| is not specified, throw a "{{TypeError}}" {{DOMException}}.
+ 1. If |version| is not `1`, throw a "{{TypeError}}" {{DOMException}}.
+ 1. If |operation| is not specified, throw a "{{TypeError}}" {{DOMException}}.
+ 1. If |operation| is not one of {{OperationType}}, throw a "{{TypeError}}" {{DOMException}}.
```
This should also not be necessary, see https://webidl.spec.whatwg.org/#es-enumeration

johannhof replied:
However, it's not clear why this should be the only check on operation, i.e. is it valid to include `operation: "token-request"` on a redemption request? Probably not.

I suspect that these types are actually used at the start (i.e. in https://wicg.github.io/trust-token-api/#http-network-or-cache-fetch) to decide which operation to perform. Currently I think the spec would tell implementers to just do both.

aykutbulut replied:
I do not think we need to add token type and version to the request. We already know them since they can only take one value. However, we do need to add the operation. We process the request based on the value of the operation field. 

1. https://source.chromium.org/chromium/chromium/src/+/main:services/network/url_loader_factory.cc;drc=089c96fe3d83e67d9a4fcc87241bad6a954ff143;l=273
2. https://source.chromium.org/chromium/chromium/src/+/main:services/network/trust_tokens/trust_token_request_helper_factory.cc;drc=bec0ef46974cd9a6bc0f1f1ae34baa11913f77af;l=140

johannhof replied:
You're saying that you want those to be validated purely through the WebIDL constraints when they're passed in as part of the privateToken argument? That seems okay for now, but yeah, you'll still have to decide what to do based on the operation.

aykutbulut replied:
Yes, verifying them through WebIDL constraints are good enough. I will add how operations and issuers are used. I think it might be better to add them in a different PR. Title of this one is specific.

johannhof reviewed (changes requested):
Thank you, this a good step forward, however, you're accessing these variables in the wrong place right now (in the network fetch vs. the fetch constructor). They should be accessed similar to the refreshPolicy in https://wicg.github.io/trust-token-api/#fetch-request, instead. For some, you may need to define new fields on `request` to carry them over in the fetch algorithm, like we do with [`token refresh policy`](https://wicg.github.io/trust-token-api/#request-token-refresh-policy). I can imagine this is needed for `operation`, at least.

johannhof reviewed (approved):
Specifying how operation works in another PR seems fine to me :)
